src/engine/Operation.cpp

```diff
@@ -48,12 +48,24 @@ vector<string> Operation::collectWarnings() const {
 }
 
 // ________________________________________________________________________
-void Operation::recursivelySetTimeoutTimer(
-    const ad_utility::SharedConcurrentTimeoutTimer& timer) {
-  _timeoutTimer = timer;
+void Operation::recursivelySetAbortionHandle(
+    SharedAbortionHandle abortionHandle) {
   for (auto child : getChildren()) {
     if (child) {
-      child->recursivelySetTimeoutTimer(timer);
+      child->getRootOperation()->recursivelySetAbortionHandle(abortionHandle);
+    }
+  }
+  abortionHandle_ = std::move(abortionHandle);
+}
+
+// ________________________________________________________________________
+
+void Operation::recursivelySetTimeConstraint(
+    std::chrono::steady_clock::time_point deadline) {
+  deadline_ = deadline;
+  for (auto child : getChildren()) {
+    if (child) {
+      child->getRootOperation()->recursivelySetTimeConstraint(deadline);
```

Comment:
Some operations "cheat" a bit when it comes to timeouts. Because they want to perform an expensive operation without intermediate cancellation checks they try to guess if the duration will take longer (estimate) than the remaining time.
However my design goal that tries to clearly abstract away the cancellation/abortion/timeout reason, which kind of clashes a bit with this.